design/baremetalhost-v1alpha2.md

```diff
@@ -0,0 +1,236 @@
+
+The CPU speed is simply being reported, and not used by the
+`baremetal-operator` code. We could, therefore, use a `string` field
+without losing any functionality. The value we get from `gophercloud`
+is already a string.
+
+Since the fractional portion of the speed can be dropped, we can use
+an `int32` to hold the value. An integer value will make it easier for
+the `hardware-classification-controller` to continue to use the value
+when applying rules.
+
+The best approach is to use an `int32` field.
+
+### Converting Existing Data
+
+It is possible to [use a web
+hook](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#configure-customresourcedefinition-to-use-conversion-webhooks)
+to convert resources from their storage format to a specific API
+format. We do not want users to continue to consume the `v1alpha1`
+format, however, because we want to be able to drop support
+quickly. Therefore, we should not write a conversion web hook.
+
+Instead, we should add startup code to the baremetal-operator to read
+all host resources in their current format and write them (with type
+conversion but no other changes) to the new storage format. See option
+2
+[here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version)
+for more discussion.
+
+### Dropping Hardware Profile Field
+
+Now that we have support for more verbose root device hints and we
+have the `hardware-classification-controller`, we no longer need the
+hardware profile fields. We can remove them as part of the
+backwards-incompatible changes as we move to the new API version.
+
+### Risks and Mitigations
+
+We do not know all of the ways that consumers of the API are using the
+CPU speed values currently being reported. Given that most kubernetes
+tools appear to assume that floating point types are not to be used in
+APIs, it is likely that most consumers are either ignoring the value
+or using it in simple ways (displaying it, logging it, etc.) where
+loss of precision from the type conversion to a float is not
+significant.
+
+### Work Items
+
+- Release `baremetal-operator` version `0.1.0` with existing API.
+- Add new `v1alpha2` API
+- Change the `ClockSpeed` type to `int32` in `v1alpha2`.
+- Remove hardware profile from spec and status.
+- Remove the state machine states for setting the hardware profile.
+- Set `v1alpha2` as the storage version
+- Add conversion code to the `baremetal-operator`
+- Update the inspection code to no longer convert the value given by
+  Ironic Inspector before storing it.
+- Release `baremetal-operator` version `0.2.0` with code that can
+  upgrade existing resources.
+- Update the module references in `cluster-api-provider-metal3` and
+  `hardware-classification-controller` and release them.
+- Drop support for `v1alpha1`
+  - Remove the API code
+  - Remove the conversion code
```

Comment:
Does this assume that anyone on 0.1.0 will upgrade to 0.2.0 on their way to 0.3.0? What would happen if someone upgrades from 0.1.0 -> 0.3.0 directly? If that needs to be prevented, how would that be done?

Reply:
From CAPI point of view, I don't think there is any way to force this. We can add it to the readme and release notes, but using clusterctl, it would be very easy to overlook this, as the command by default pulls the latest version

Reply:
I also think it'll be difficult to guarantee that people upgrade to 0.2.0.

Once the controller itself is using v1alpha2, then all we need the v1alpha1 structs for is the conversion webhook, right? Maybe we can "freeze" that code and have operator-sdk stop interacting with it during `operator-sdk generate` or similar. Moving the v1alpha1 code to a different package might be sufficient for it to be ignored.

Reply:
The tool only includes information in the generated CRD for versions that it reads.